### PR TITLE
Add method NodeBox::into_value.

### PR DESCRIPTION
**Pull request by efenniht**
Make a method to unbox `NodeBox` containing `NodeBodyV`. Suggests a canonical way for kaist-cp/helpdesk#28 and kaist-cp/helpdesk#40.

### CODE REVIEW

File: art/src/node.rs [hunk 1]
```diff
@@ -476,6 +476,22 @@ impl<V> NodeBox<V> {
...
+    ///
+    /// # Panics
+    ///
+    /// Panics if it does not contain `NodeBodyV`.
+    pub fn into_value(self) -> V {
+        let ptr = self.inner & !TAG_MASK;
+        assert_eq!(self.inner & TAG_MASK, 4);
+
+        let node = unsafe { Box::from_raw(ptr as *mut CachePadded<(NodeHeader, NodeBodyV<V>)>) };
+        mem::forget(self);
```
Comment (tomtomjhj) [inline art/src/node.rs]: why is `forget` needed here?

Comment (efenniht) [reply art/src/node.rs]: If there is no `forget`, `self` is dropped and it causes double free (first free occurs `*node` at L492)

Comment (tomtomjhj) [reply art/src/node.rs]: This is what valgrind says if I remove that `forget`.

```
Invalid free() / delete / delete[] / realloc()
   at 0x483997B: free (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
   by 0x1504D2: alloc::alloc::dealloc (alloc.rs:106)
   by 0x15B731: alloc::alloc::box_free (alloc.rs:224)
   by 0x15A71A: core::ptr::real_drop_in_place (mod.rs:175)
   by 0x162E3C: core::mem::drop (mod.rs:708)
   by 0x136BC1: NodeBox<V>::drop_inner (node.rs:413)
   by 0x13B9D8: <NodeBox<V> as core::ops::drop::Drop>::drop (node.rs:521)
   by 0x159C1D: core::ptr::real_drop_in_place (mod.rs:175)
   by 0x137262: NodeBox<V>::into_value (node.rs:396)
   by 0x153652: Entry<V,I>::delete (art.rs:141)
   by 0x1505C6: <Art<V> as SequentialMap<V>>::delete (art.rs:249)
   by 0x130A11: art_test::smoke (art_test.rs:34)
 Address 0x4bf2e00 is 0 bytes inside a block of size 128 free'd
   at 0x483997B: free (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
   by 0x1504D2: alloc::alloc::dealloc (alloc.rs:106)
   by 0x15B731: alloc::alloc::box_free (alloc.rs:224)
   by 0x137255: NodeBox<V>::into_value (node.rs:396)
   by 0x153652: Entry<V,I>::delete (art.rs:141)
   by 0x1505C6: <Art<V> as SequentialMap<V>>::delete (art.rs:249)
   by 0x130A11: art_test::smoke (art_test.rs:34)
   by 0x127029: art_test::smoke::{{closure}} (art_test.rs:22)
   by 0x15974D: core::ops::function::FnOnce::call_once (function.rs:227)
   by 0x16F65E: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once (boxed.rs:910)
   by 0x216229: __rust_maybe_catch_panic (lib.rs:80)
   by 0x189C82: try<(),std::panic::AssertUnwindSafe<alloc::boxed::Box<FnOnce<()>>>> (panicking.rs:271)
   by 0x189C82: catch_unwind<std::panic::AssertUnwindSafe<alloc::boxed::Box<FnOnce<()>>>,()> (panic.rs:394)
   by 0x189C82: run_test_in_process (lib.rs:1626)
   by 0x189C82: test::run_test::run_test_inner::{{closure}} (lib.rs:1504)
 Block was alloc'd at
   at 0x483AE93: memalign (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
   by 0x483AFC0: posix_memalign (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
   by 0x20E679: aligned_malloc (alloc.rs:91)
   by 0x20E679: alloc (alloc.rs:22)
   by 0x20E679: __rdl_alloc (alloc.rs:239)
   by 0x15046B: alloc::alloc::alloc (alloc.rs:84)
   by 0x1503DB: alloc::alloc::exchange_malloc (alloc.rs:206)
   by 0x139ECA: new<crossbeam_utils::cache_padded::CachePadded<(NodeHeader, NodeBodyV<i32>)>> (boxed.rs:121)
   by 0x139ECA: NodeBox<V>::new_inner (node.rs:399)
   by 0x138454: NodeBox<V>::newv (node.rs:448)
   by 0x138BF4: NodeBox<V>::new_path (node.rs:470)
   by 0x152A47: Entry<V,I>::or_insert_with (art.rs:70)
   by 0x153B82: Entry<V,I>::or_insert (art.rs:113)
   by 0x150747: <Art<V> as SequentialMap<V>>::insert (art.rs:244)
   by 0x1307F2: art_test::smoke (art_test.rs:29)
```
So the thing being double free'd is `CachePadded<(NodeHeader, NodeBodyV<i32>)>`. I don't get why it gets dropped when `self` is dropped.

Comment (efenniht) [reply art/src/node.rs]: Valgrind says:
```
Invalid free() / delete / delete[] / realloc()
...
   by 0x136BC1: NodeBox<V>::drop_inner (node.rs:413)
   by 0x13B9D8: <NodeBox<V> as core::ops::drop::Drop>::drop (node.rs:521)
```

So the destructor of `NodeBox` drops its inner `NodeBody`.

Comment (efenniht) [reply art/src/node.rs]: See https://github.com/kaist-cp/cs492-concur/pull/15/files/033b0e48746335564d8fae58767e04cfea411121#diff-4d72c780525dbb9c8c9d668ba3359fc2R521 here.

Comment (tomtomjhj) [reply art/src/node.rs]: ohhh thanks

File: art/src/node.rs [hunk 2]
```diff
@@ -476,6 +476,22 @@ impl<V> NodeBox<V> {
...
     }
+
+    /// Unboxes `self` and returns its containing value.
+    ///
+    /// # Panics
+    ///
+    /// Panics if it does not contain `NodeBodyV`.
+    pub fn into_value(self) -> V {
+        let ptr = self.inner & !TAG_MASK;
+        assert_eq!(self.inner & TAG_MASK, 4);
```
Comment (jeehoonkang) [inline art/src/node.rs]: How about let it return `Option<V>` rather than asserting here?

Comment (efenniht) [reply art/src/node.rs]: This code is designed to follow the convention of `NodeBox::into` -- it panics if `self` does not contain a kind of `NodeBodyI`. I agree that `Option<V>` is better but if so, APIs are a bit less consistent, IMO.

**Pull request merged**